Tests

martinheidegger opened the pull request:
In #7 You asked why there was a null result. Instead of trying to explain it I added some simple test case that tries out how that command is executed against the graphql API and then uses a dummy db to make sure that the right request to the database is executed. 

Notes: 
- `t.deepEquals` is one of [node-tap's assert functions](http://www.node-tap.org/asserts/)
- `npm test` runs the tests.
- I helped to seperate the Schema creation so I moved it to `initSchema`

martinheidegger commented on ql/loaders.js:
```diff
@@ -15,8 +15,7 @@ module.exports = (db) => {
 				})
-				.then(rows => console.log(rows))
-				.catch(err => console.error(err))
+				.then(rows => rows)
```
This was the reason for the missing user return

martinheidegger commented on test/db.js:
```diff
@@ -0,0 +1,53 @@
+    select(fields) {
+      // Immitate the real database API used
+      const p = new Promise(resolve => setImmediate(() => {
+        t.deepEquals(props.from, ['users'], 'selecting a user')
```
This test is good for figuring out how Graphql works and if Graphql is setup properly but it is a bad idea to use this as a template for future tests

martinheidegger commented on test/db.js:
```diff
@@ -0,0 +1,53 @@
+      }
+    }
+  })
+})
```
Argh, I forgot the line at the end of the file

martinheidegger commented on server.js:
```diff
@@ -7,99 +7,24 @@ const next = require('next');
-				query: RootQuery,
-				mutation: RootMutation
-			})
+			const Schema = initSchema(db);
```
This PR just reduces the schema but actually the server setup might become a lot smaller if you want to add future tests

vtrofin commented:
Your answer opened some more questions. I will start with the questions arising after going through the test code (i am struggling a bit to understand the addDummyAPI function and the const schema entirely).

1. How did you realize, based on running the test, that me console.logging the response  in the getNodeById function (after returning the response initially) was the issue in the first place? Because I can't figure this out based on the test, even after running it.

2. How do I improve error reporting/ debugging between nodejs and graphql so I always know where I should look for a culprit? The “data: null” response in GraphiQL and lack of error messages didn’t make me think at all that getNodeById function is the issue. I can only think of your example from the last issue (central knex config) 

```javascript
let sensitive;
 try {
   sensitive = require('./ql/sensitive');
 } catch (e) {
   console.warn('Missing sensitive knex configuration!');
   sensitive = {};
 }
```
Is there anything I should use on top of that?

martinheidegger commented:
> (i am struggling a bit to understand the addDummyAPI function and the const schema entirely)

`addDummyAPI` was the quickest and most-straightforward way I found to implement the knex API in a way that is allows me to make sure all calls are performed and I was able to monitor what arguments were used.

> 1. How did you realize, based on running the test, that me console.logging the response in the getNodeById function (after returning the response initially) was the issue in the first place? Because I can't figure this out based on the test, even after running it.

Well, tbh. I read the mistake while reading the code and saw the mistake on first glance. However I tried to ignore what I saw and implemented the test. It turned out that with the dummy running, the response was null. So I looked in places between the knex api and the graphql api that could "eat" something and found the Promises that omit things and found the `.then` statement that only logged but didn't return anything, so I fixed that and voila - it worked.

> 2. How do I improve error reporting/ debugging between nodejs and graphql so I always know where I should look for a culprit? The “data: null” response in GraphiQL and lack of error messages didn’t make me think at all that getNodeById function is the issue. I can only think of your example from the last issue (central knex config)

Hmm, the problem is that `data:null` is not an invalid response. Its technically an okay response that is indistinguishable from the error response. So there are two things i can think of right of the top of my hat to make this less of an issue:

1. Add a thin "layer" between GraphQL that logs all requests from GraphQL to your implementation. This way you know what things have been sent to your API and what things your implementation returned.
2. Treat empty results as error cases.
    ```javascript
     .then(rows => {
         if(!rows || rows.length === 0) return Promise.reject(new Error('not-found'))
         // ...
      })
     ```
